### PR TITLE
Kernel: Harden construction patterns against OOM conditions

### PR DESCRIPTION
**Pull request by supercomputer7**
Relies on #9818 and #9871 (to some extent).
Fixes problems that were mentioned in https://github.com/SerenityOS/serenity/commit/75564b4a5f2b5452163571116ee9efaf5c3c65af.

### CODE REVIEW

File: Kernel/Bus/VirtIO/Console.cpp [hunk 1]
```diff
@@ -12,14 +12,24 @@ namespace Kernel::VirtIO {
...
 
-UNMAP_AFTER_INIT NonnullRefPtr<Console> Console::must_create(PCI::Address address)
+UNMAP_AFTER_INIT RefPtr<Console> Console::try_create(PCI::Address address)
 {
-    return adopt_ref_if_nonnull(new Console(address)).release_nonnull();
+    auto control_receive_buffer = Memory::RingBuffer::try_create("VirtIOConsole control receive queue", CONTROL_BUFFER_SIZE);
+    auto control_transmit_buffer = Memory::RingBuffer::try_create("VirtIOConsole control transmit queue", CONTROL_BUFFER_SIZE);
+    if ((!control_receive_buffer.is_error()) || (control_transmit_buffer.is_error()))
+        return {};
+    auto console = adopt_ref_if_nonnull(new Console(address, control_receive_buffer.release_value(), control_transmit_buffer.release_value()));
```
Comment (Dexesttp) [inline Kernel/Bus/VirtIO/Console.cpp:21]:
```suggestion
    auto console = adopt_ref_if_nonnull(new (nothrow) Console(address, control_receive_buffer.release_value(), control_transmit_buffer.release_value()));
```

File: Kernel/Bus/VirtIO/RNG.cpp [hunk 2]
```diff
@@ -9,14 +9,23 @@
...
 namespace Kernel::VirtIO {
 
-UNMAP_AFTER_INIT NonnullRefPtr<RNG> RNG::must_create(PCI::Address address)
+UNMAP_AFTER_INIT RefPtr<RNG> RNG::try_create(PCI::Address address)
 {
-    return adopt_ref_if_nonnull(new RNG(address)).release_nonnull();
+    auto entropy_buffer = MM.allocate_contiguous_kernel_region(PAGE_SIZE, "VirtIO::RNG", Memory::Region::Access::ReadWrite);
+    if (entropy_buffer.is_error())
+        return {};
+    auto rng = adopt_ref_if_nonnull(new RNG(address, entropy_buffer.release_value()));
```
Comment (Dexesttp) [inline Kernel/Bus/VirtIO/RNG.cpp:17]:
```suggestion
    auto rng = adopt_ref_if_nonnull(new (nothrow) RNG(address, entropy_buffer.release_value()));
```

File: Kernel/Graphics/VirtIOGPU/GPU.cpp [hunk 3]
```diff
@@ -44,15 +46,26 @@ void GPU::initialize()
...
     }
+    return true;
+}
+
+UNMAP_AFTER_INIT RefPtr<GPU> GPU::try_create(Badge<VirtIOGPU::GraphicsAdapter>, PCI::Address address)
+{
+    auto scratch_region = MM.allocate_contiguous_kernel_region(32 * PAGE_SIZE, "VirtGPU Scratch Space", Memory::Region::Access::ReadWrite);
+    if (scratch_region.is_error())
+        return {};
+    auto gpu_device = adopt_ref_if_nonnull(new GPU(address, scratch_region.release_value()));
```
Comment (Dexesttp) [inline Kernel/Graphics/VirtIOGPU/GPU.cpp:57]:
```suggestion
    auto gpu_device = adopt_ref_if_nonnull(new (nothrow) GPU(address, scratch_region.release_value()));
```

File: Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp [hunk 4]
```diff
@@ -12,17 +12,27 @@
 
 namespace Kernel::Graphics::VirtIOGPU {
 
-NonnullRefPtr<GraphicsAdapter> GraphicsAdapter::initialize(PCI::Address base_address)
+RefPtr<GraphicsAdapter> GraphicsAdapter::initialize(PCI::Address base_address)
 {
     VERIFY(PCI::get_id(base_address).vendor_id == PCI::VendorID::VirtIO);
-    return adopt_ref(*new GraphicsAdapter(base_address));
+    auto adapter = adopt_ref_if_nonnull(new GraphicsAdapter(base_address));
```
Comment (Dexesttp) [inline Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp:18]:
```suggestion
    auto adapter = adopt_ref_if_nonnull(new (nothrow) GraphicsAdapter(base_address));
```

File: Kernel/Storage/AHCIController.cpp [hunk 5]
```diff
@@ -14,9 +14,15 @@
 
 namespace Kernel {
 
-NonnullRefPtr<AHCIController> AHCIController::initialize(PCI::Address address)
+KResultOr<NonnullRefPtr<AHCIController>> AHCIController::try_create(PCI::Address address)
 {
-    return adopt_ref(*new AHCIController(address));
+    auto hba_region = TRY(MM.allocate_kernel_region(PhysicalAddress(PCI::get_BAR5(address)).page_base(), Memory::page_round_up(sizeof(AHCI::HBA)), "AHCI HBA", Memory::Region::Access::ReadWrite));
+    auto controller = adopt_ref_if_nonnull(new AHCIController(move(hba_region), address));
```
Comment (Dexesttp) [inline Kernel/Storage/AHCIController.cpp:20]:
```suggestion
    auto controller = adopt_ref_if_nonnull(new (nothrow) AHCIController(move(hba_region), address));
```

File: Kernel/Storage/AHCIPort.cpp [hunk 6]
```diff
@@ -19,9 +19,36 @@
...
+    return (registers.ssts & 0xf) == 4;
+}
+
+KResultOr<NonnullRefPtr<AHCIPort>> AHCIPort::try_create(const AHCIPortHandler& handler, volatile AHCI::PortRegisters& registers, u32 port_index)
+{
+    if (is_interface_disabled(registers)) {
+        auto port = adopt_ref_if_nonnull(new AHCIPort(handler, registers, port_index));
+        if (!port)
+            return KResult(ENOMEM);
+        return port.release_nonnull();
```
Comment (Dexesttp) [inline Kernel/Storage/AHCIPort.cpp:33]: This is such a common pattern that this now exists :) (also there's the extra `(nothrow)`)
```suggestion
        return adopt_nonnull_ref_or_enomem(new (nothrow) AHCIPort(handler, registers, port_index));
```

File: Kernel/Storage/AHCIPort.cpp [hunk 7]
```diff
@@ -19,9 +19,36 @@
...
+    if (dma_buffer_page.is_null())
+        return ENOMEM;
+    RefPtr<Memory::PhysicalPage> command_table_page = MM.allocate_supervisor_physical_page();
+    if (command_table_page.is_null())
+        return ENOMEM;
+    auto command_list_region = TRY(MM.allocate_kernel_region(command_list_page->paddr(), PAGE_SIZE, "AHCI Port Command List", Memory::Region::Access::ReadWrite, Memory::Region::Cacheable::No));
+    auto port = adopt_ref_if_nonnull(new AHCIPort(handler, registers, move(command_list_region), *command_list_page, *fis_receive_page, *dma_buffer_page, *command_table_page, port_index));
+    if (!port)
+        return KResult(ENOMEM);
+    return port.release_nonnull();
```
Comment (Dexesttp) [inline Kernel/Storage/AHCIPort.cpp:51]:
```suggestion
    return adopt_nonnull_ref_or_enomem(new (nothrow) AHCIPort(handler, registers, move(command_list_region), *command_list_page, *fis_receive_page, *dma_buffer_page, *command_table_page, port_index));
```

File: Kernel/Storage/AHCIPortHandler.cpp [hunk 8]
```diff
@@ -9,41 +9,55 @@
...
     for (size_t index = 0; index < (((size_t)AHCI::Limits::MaxPorts * 512) / PAGE_SIZE); index++) {
-        m_identify_metadata_pages.append(MM.allocate_supervisor_physical_page().release_nonnull());
+        auto page = MM.allocate_supervisor_physical_page();
+        if (!page)
+            return KResult(ENOMEM);
+        identify_metadata_pages.append(page.release_nonnull());
     }
 
-    dbgln_if(AHCI_DEBUG, "AHCI Port Handler: IRQ {}", irq);
+    auto handler = adopt_ref_if_nonnull(new AHCIPortHandler(controller, irq, identify_metadata_pages, taken_ports));
```
Comment (Dexesttp) [inline Kernel/Storage/AHCIPortHandler.cpp:23]:
```suggestion
    auto handler = adopt_ref_if_nonnull(new (nothrow) AHCIPortHandler(controller, irq, identify_metadata_pages, taken_ports));
```

File: Kernel/Memory/RingBuffer.cpp [hunk 9]
```diff
@@ -4,14 +4,24 @@
...
 
-RingBuffer::RingBuffer(String region_name, size_t capacity)
-    : m_region(MM.allocate_contiguous_kernel_region(page_round_up(capacity), move(region_name), Region::Access::Read | Region::Access::Write).release_value())
+KResultOr<NonnullOwnPtr<RingBuffer>> RingBuffer::try_create(String region_name, size_t capacity)
+{
+    auto region = TRY(MM.allocate_contiguous_kernel_region(page_round_up(capacity), move(region_name), Region::Access::Read | Region::Access::Write));
+    auto ring_buffer = adopt_own_if_nonnull(new RingBuffer(move(region), capacity));
+    if (!ring_buffer)
+        return ENOMEM;
+    return ring_buffer.release_nonnull();
```
Comment (Dexesttp) [inline Kernel/Memory/RingBuffer.cpp:20]: This also exists in an `own` flavor, so might as well :^)
```suggestion
    return adopt_nonnull_own_or_enomem(new (nothrow) RingBuffer(move(region), capacity));
```

**Review by Dexesttp — COMMENTED** [review]
I know it's still a draft, but I was idly reading through this (to stay up to date on these changes) and I noticed a few missing `(nothrow)`. Also, some `KError(ENOMEM)` patterns that were made easier very recently with some new helpers.

So here's a random review!

**Comment by supercomputer7** [comment]
There's plenty of work to be done in other current active PRs, I figured I should temporarily close this.

**Pull request closed without merging**